optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,344 @@
+            "if this independent sampling is intended behavior."
+        )
+
+    def sample_independent(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+
+        if self._initial_search_space is not None:
+            if self._warn_independent_sampling:
+                self._log_independent_sampling(trial, param_name)
+
+        return self._independent_sampler.sample_independent(
+            study, trial, param_name, param_distribution
+        )
+
+    def sample_relative(
+        self, study: Study, trial: FrozenTrial, search_space: Dict[str, BaseDistribution]
+    ) -> Dict[str, Any]:
+
+        if search_space == {}:
+            return {}
+
+        sample = self._sample_qmc(study, search_space)
+        trans = _SearchSpaceTransform(search_space)
+        sample = trans.bounds[:, 0] + sample * (trans.bounds[:, 1] - trans.bounds[:, 0])
+        return trans.untransform(sample[0, :])
+
+    def after_trial(
+        self,
+        study: "optuna.Study",
+        trial: "optuna.trial.FrozenTrial",
+        state: TrialState,
+        values: Optional[Sequence[float]],
+    ) -> None:
+        self._independent_sampler.after_trial(study, trial, state, values)
+
+    def _sample_qmc(
+        self, study: Study, search_space: Dict[str, BaseDistribution]
+    ) -> numpy.ndarray:
+
+        # Lazy import because the `scipy.stats.qmc` is slow to import.
+        import scipy.stats.qmc
+
+        sample_id = self._find_sample_id(study, search_space)
+        d = len(search_space)
+
+        # Use cached `qmc_engine` or construct a new one.
+        if self._is_engine_cached(d, sample_id):
+            qmc_engine: scipy.stats.qmc.QMCEngine = self._cached_qmc_engine
+        else:
+            if self._qmc_type == "halton":
+                qmc_engine = scipy.stats.qmc.Halton(d, seed=self._seed, scramble=self._scramble)
+            elif self._qmc_type == "sobol":
+                qmc_engine = scipy.stats.qmc.Sobol(d, seed=self._seed, scramble=self._scramble)
+            else:
+                message = (
+                    f"The `qmc_type`, {self._qmc_type}, is not a valid. "
+                    'It must be one of "halton" and "sobol".'
+                )
+                raise ValueError(message)
+
+        forward_size = sample_id - qmc_engine.num_generated  # `sample_id` starts from 0.
+        qmc_engine.fast_forward(forward_size)
+        sample = qmc_engine.random(1)
+        self._cached_qmc_engine = qmc_engine
+
+        return sample
+
+    def _find_sample_id(self, study: Study, search_space: Dict[str, BaseDistribution]) -> int:
+
+        qmc_id = ""
+        qmc_id += self._qmc_type
+        qmc_id += str(search_space)
+        # Sobol/Halton sequences without scrambling do not use seed.
+        if self._scramble:
+            qmc_id += str(self._seed)
+        hashed_qmc_id = hash(qmc_id)
```

Comment:
[question] Sorry, but why do we need hash keys from `seed` and `search_space`?

Reply:
I intended to shorten the length of `qmc_id` but the use of `hash()` is not a very strong preference. As far as I can remember, the length of `qmc_id` was longer at first as it supported more kinds of QMC sequences so there was more incentive to shorten it when I coded this. If you think hashing is unnecessary, please let me know. I can just remove this.